### PR TITLE
Add addon jest

### PR DESCRIPTION
**Pull request by renaudtertrais**
Issue:

## What I did

Add addon-jest to monorepo

## How to test

In storybook example

### CODE REVIEW

File: addons/jest/.eslintignore [hunk 1]
```diff
@@ -0,0 +1,2 @@
+node_modules
+dist
```
Comment (Hypnosphi) [inline addons/jest/.eslintignore]: Those should be covered by root .eslintignore file

File: addons/jest/.eslintrc.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+  },
+  "settings": {
+    "import/resolver": {
+      "webpack": {
+        config: 'webpack-development.config.js'
+      },
+    },
+  },
+};
+
```
Comment (Hypnosphi) [inline addons/jest/.eslintrc.js]: Let's try to use our root eslint config for this package, and override only the things that really need it, if any

File: addons/jest/example/.storybook/config.js [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+  url: 'https://github.com/storybooks/storybook',
+  downPanelInRight: true,
+  showLeftPanel: true,
+});
+
+function loadStories() {
+  require('../List.story');
+}
+
+configure(loadStories, module);
```
Comment (Hypnosphi) [inline addons/jest/example/.storybook/config.js]: Let's move the stories to `example/cra-kitchen-sink`

**Comment by Hypnosphi** [comment]
@shilman is a new adding considered a "small feature"?

**Review by Hypnosphi — CHANGES_REQUESTED** [review]
yarn.lock should be removed as well: we use yarn workspaces feature, so all the dependencies are installed by root `yarn bootstrap` command

File: addons/jest/package.json [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    "register.js"
+  ],
+  "main": "dist/index.js",
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/storybooks/storybook"
+  },
+  "scripts": {
+    "prebuild": "npm run clear",
+    "build": "cross-env NODE_ENV=production babel -d ./dist ./src",
```
Comment (Hypnosphi) [inline addons/jest/package.json]: We need a `prepare` script instead (see other addons). All the babel-related stuff can be removed from package as we use top-level babel dependencies and setup for transpiling everything

**Review by Hypnosphi — CHANGES_REQUESTED** [review]

File: addons/jest/example/.storybook/config.js (hunk 3 above)
Comment (ndelangen) [reply addons/jest/example/.storybook/config.js]: agreed

File: addons/jest/package.json [hunk 5]
```diff
@@ -56,16 +41,14 @@
...
     "eslint-plugin-react": "^7.4.0",
-    "gh-pages": "^1.0.0",
     "jest": "^21.2.1",
     "react": "^16.1.0",
     "react-dom": "^16.1.0",
     "style-loader": "^0.19.0"
   },
   "peerDependencies": {
-    "@storybook/addons": "^3.2.14",
     "prop-types": "^15.6.0",
```
Comment (Hypnosphi) [inline addons/jest/package.json]: It should be a simple dependency, not a peer one

**Review by Hypnosphi — CHANGES_REQUESTED** [review]

File: addons/jest/package.json (hunk 5 above)
Comment (Hypnosphi) [reply addons/jest/package.json]: I actually meant the `prop-types` package. It's OK for an end user not to have it

**Comment by ndelangen** [comment]
Ready to merge I think? @Hypnosphi ?

**Comment by Hypnosphi** [comment]
Please move `prop-types` to dependencies

**Review by Hypnosphi — APPROVED** [review]

File: addons/jest/example/.storybook/config.js (hunk 3 above)
Comment (renaudtertrais) [reply addons/jest/example/.storybook/config.js]: I am on it.

File: addons/jest/package.json [hunk 6]
```diff
@@ -26,7 +26,8 @@
   "dependencies": {
     "@storybook/addons": "^3.2.15",
     "global": "^4.3.2",
-    "prop-types": "^15.6.0"
+    "prop-types": "^15.6.0",
+    "styled-components": "^2.2.3"
```
Comment (Hypnosphi) [inline addons/jest/package.json]: @norbert Are we OK with having `glamorous`, `style-components`, inline styles, and even some good old css in the same project?

File: addons/jest/README.md [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+```
+
+## Usage
+
+Assuming that you have created a test files `MyComponent.test.js` and `MyOtherComponent.test.js`
+
+In your `story.js`
+
+```js
+import jestTestResults from '../.jest-test-results.json';
```
Comment (Hypnosphi) [inline addons/jest/README.md]: Does that mean that you have to run the tests before ever starting storybook?

File: addons/jest/package.json [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+  "main": "dist/index.js",
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/storybooks/storybook"
+  },
+  "scripts": {
+    "prepare": "node ../../scripts/prepare.js"
+  },
+  "dependencies": {
+    "@storybook/addons": "^3.2.15",
```
Comment (Hypnosphi) [inline addons/jest/package.json]: Let's make it a peerDep from day one

File: examples/cra-kitchen-sink/src/components/SimpleList.js [hunk 9]
```diff
@@ -0,0 +1,14 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+const List = ({ items }) => (
+  <ul style={{ fontFamily: 'Helvetica, Arial, sans-srif', fontSize: 16 }}>
```
Comment (Hypnosphi) [inline examples/cra-kitchen-sink/src/components/SimpleList.js]: `srif` must be a typo

File: addons/jest/src/components/Panel.js [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+
+import Indicator from './Indicator';
+import Result from './Result';
+import provideJestResult from '../hoc/provideJestResult';
+import colors from '../colors';
+
+const Panel = ({ tests }) => {
+  const style = {
+    padding: '10px 20px',
+    flex: 1,
```
Comment (Hypnosphi) [inline addons/jest/src/components/Panel.js]: Let's add our baseFonts here

**Review by Hypnosphi — CHANGES_REQUESTED** [review]

File: addons/jest/src/components/Panel.js (hunk 10 above)
Comment (Hypnosphi) [reply addons/jest/src/components/Panel.js]: Another option which I actually like more is to add it to top-level `manager` [element](https://github.com/storybooks/storybook/blob/master/lib/ui/src/modules/ui/components/layout/index.js#L9)

**Comment by ndelangen** [comment]
Latest screenshot:
screen shot 2017-11-19 at 00 34 56

**Comment by Hypnosphi** [comment]
Actually, I still don't get what's happening here =( Why those green and red bars have different length? Why storybook is a place to show test results in a fancy way? Why would anybody want to make empty stories?

File: addons/jest/README.md [hunk 11]
```diff
@@ -0,0 +1,133 @@
+# Storybook addon Jest
+
+Brings Jest results in storybook.
+
+[![Storybook Jest Addon Demo](@storybook/addon-jest.gif)](https://@storybook/addon-jest-example.herokuapp.com/)
```
Comment (Hypnosphi) [inline addons/jest/README.md]: The link is broken

File: addons/jest/README.md (hunk 7 above)
Comment (ndelangen) [reply addons/jest/README.md]: Yes, edited the readme to make this clear.

For demo purposes I've added the testresults-file to source control.

**Review by shilman — CHANGES_REQUESTED** [review]
What's up with the `SimpleList` CI failures? Looks like some files didn't get checked in?

Aside from that, this looks great. @Hypnosphi this is a new feature so technically it should probably go into 3.3, but it's isolated and shouldn't break anything so I'm also OK with it to go into a patch.

Very excited to get this into peoples' hands in one way or another and look forward to getting a cleaned up version merged soon after the changes have been addressed.

**Comment by Hypnosphi** [comment]
> it's isolated and shouldn't break anything

Actually isolation is an issue here. Once someone adds this addon for some of the stories, nobody can `start-storybook` in that project without running those tests locally first.

File: addons/jest/README.md (hunk 7 above)
Comment (Hypnosphi) [reply addons/jest/README.md]: > For demo purposes

What exactly does it demonstrate? Users are not supposed to create those files manually, are they?

File: addons/jest/README.md [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+
+`package.json`
+
+```js
+"scripts": {
+  "test": "jest --json --outputFile=.jest-test-results.json"
+}
+```
+
+Add it the result file to `.gitignore`:
```
Comment (Hypnosphi) [inline addons/jest/README.md]: @renaudtertrais @ndelangen why do we even have this recommendation? Checking it into VCS makes a lot of sense to me

Comment (ndelangen) [reply addons/jest/README.md]: Well, it will likely have tons of merge-conflicts over time?

Comment (Hypnosphi) [reply addons/jest/README.md]: Just like lockfiles and test snapshots. It's easy to regenerate them, so this shouldn't be a problem

Comment (ndelangen) [reply addons/jest/README.md]:
```js
true
```

I will change the text and make it so users are aware of the trade-offs.

**Comment by ndelangen** [comment]
@shilman, @Hypnosphi I think I've resolved all issues? Can you accept?

**Review by Hypnosphi — APPROVED** [review]

**Comment by ndelangen** [comment]
@shilman The thing you mentioned has been resolved, I hope you're OK with me merging this.

**Pull request merged**